lib/recurly.d.ts

```diff
@@ -0,0 +1,785 @@
+  discount?: CouponDiscount;
+  couponType?: string;
+  expiredAt?: Date;
+}
+
+export interface SubscriptionChange {
+  id?: string;
+  object?: string;
+  subscriptionId?: string;
+  plan?: PlanMini;
+  addOns?: SubscriptionAddOn[];
+  unitAmount?: number;
+  quantity?: number;
+  shipping?: SubscriptionShipping;
+  activateAt?: Date;
+  activated?: boolean;
+  createdAt?: Date;
+  updatedAt?: Date;
+  deletedAt?: Date;
+}
+
+export interface SubscriptionAddOn {
+  id?: string;
+  object?: string;
+  subscriptionId?: string;
+  addOn?: AddOnMini;
+  quantity?: number;
+  unitAmount?: number;
+  createdAt?: Date;
+  updatedAt?: Date;
+  expiredAt?: Date;
+}
+
+export interface AddOnMini {
+  id?: string;
+  object?: string;
+  code?: string;
+  name?: string;
+  accountingCode?: string;
+}
+
+export interface UniqueCouponCode {
+  id?: string;
+  object?: string;
+  code?: string;
+  state?: string;
+  createdAt?: Date;
+  updatedAt?: Date;
+  redeemedAt?: Date;
+  expiredAt?: Date;
+}
+
+export interface CustomFieldDefinition {
+  id?: string;
+  object?: string;
+  relatedType?: string;
+  name?: string;
+  userAccess?: string;
+  displayName?: string;
+  tooltip?: string;
+  createdAt?: Date;
+  updatedAt?: Date;
+  deletedAt?: Date;
+}
+
+export interface Plan {
+  id?: string;
+  object?: string;
+  code?: string;
+  state?: string;
+  name?: string;
+  description?: string;
+  intervalUnit?: string;
+  intervalLength?: number;
+  trialUnit?: string;
+  trialLength?: number;
+  totalBillingCycles?: number;
+  autoRenew?: boolean;
+  accountingCode?: string;
+  setupFeeAccountingCode?: string;
+  taxCode?: string;
+  taxExempt?: boolean;
+  currencies?: PlanPricing[];
+  hostedPages?: PlanHostedPages;
+  createdAt?: Date;
+  updatedAt?: Date;
+  deletedAt?: Date;
+}
+
+export interface PlanPricing {
+  currency?: string;
+  setupFee?: number;
+  unitAmount?: number;
+}
+
+export interface PlanHostedPages {
+  successUrl?: string;
+  cancelUrl?: string;
+  bypassConfirmation?: boolean;
+  displayQuantity?: boolean;
+}
+
+export interface AddOn {
+  id?: string;
+  object?: string;
+  planId?: string;
+  code?: string;
+  state?: string;
+  name?: string;
+  accountingCode?: string;
+  taxCode?: string;
+  displayQuantity?: boolean;
+  defaultQuantity?: number;
+  currencies?: AddOnPricing[];
+  createdAt?: Date;
+  updatedAt?: Date;
+  deletedAt?: Date;
+}
+
+export interface AddOnPricing {
+  currency?: string;
+  unitAmount?: number;
+}
+
+export interface ShippingMethod {
+  id?: string;
+  object?: string;
+  code?: string;
+  name?: string;
+  taxCode?: string;
+  createdAt?: Date;
+  updatedAt?: Date;
+  deletedAt?: Date;
+}
+
+
+
+export interface Empty {
+}
+
+export declare class Pager<T> {
+  count(): number;
+  first(): T;
+  each(): AsyncIterable<T>;
+  eachPage(): AsyncIterable<T>;
+}
+
+export declare class Client {
+  constructor(apiKey: string);
+  apiVersion(): string;
+  listSites(params: object): Pager<Site>;
```

Comment:
are you able to type the parameters to this and other methods? if you don't have time i understand, but that could be helpful.

Reply:
I think this covers all the methods on the Client class. Which were you referring to? I want to make sure it's complete.

Reply:
Oh, you mean instead of a plain object, it could be a type? It's doable with our generator. I might do that in a follow up though. I'll investigate.

Reply:
So, right now both Request bodies and parameters are untyped as that's how they are in the Javascript world. However we do provided typed versions of these for java, .net, go. It can  be added but I'm going to plan that as a follow up.

Reply:
sounds good to me. yeah my point is just that as far as typescript types are convenient for documentation and development, being able to autocomplete parameters etc would be very handy :D

Reply:
Yeah, agreed. I find our Dotnet and Java development experiences to be better because of this. These features will come soon!

Reply:
@wcauchois Has covered everything I had seen, only thing I have is that the parameter objects in these methods should be optional:
```TypeScript
listSites(params?: object): Pager<Site>;
```

Reply:
Just as an example of what I'd expect from this particular parameter object interface:
```TypeScript
interface ListSitesParameters {
       ids?: string[];
       limit?: number;
       order?: 'asc' | 'desc';
       sort?: keyof Site;
}
```

Notice the `keyof Site` - I don't know if your client converts camelcase sort parameters to snakecase as the API would expect them to be so this might not work.

Reply:
Good catch! Yes, i'll have those parameterized in the next release